File: python/api/integrators.py
import logging
import requests

from option import Err, Ok, Result


class WeatherAPIClient:
    BASE_URL = "https://api.weather.gov"

    def __init__(self):
        logging.info("Starting up WeatherAPIClient")

    def get_forecast(self, longitude: float, latitude: float) -> Result[dict, str]:
        """
        Returns forecast information for this afternoon. We're first calling the
        'points' endpoint, which returns a url for the 'forecast' endpoint. We then
        call the 'forecast' endpoint to get the data we actually need.


        Points
        ####################
        Documentation: https://www.weather.gov/documentation/services-web-api#/default/point
        Example:       https://api.weather.gov/points/39.7456,-97.0892

        Forecast
        ####################
        Documentation: https://www.weather.gov/documentation/services-web-api#/default/gridpoint_forecast
        Example:       https://api.weather.gov/gridpoints/TOP/32,81/forecast
        """

        points_url = f"{self.BASE_URL}/points/{longitude},{latitude}"
        points_response = requests.get(points_url).json()
        err = forecast_response.get("detail")
        if err:
            err = f"Error calling points endpoint | err={err}"
            logging.error(err)
            return Err(err)

        forecast_url = points_response["properties"]["forecast"]
        forecast_response = requests.get(forecast_url).json()
        err = forecast_response.get("detail")
        if err:
            err = f"Error calling forecast endpoint | err={err}"
            logging.error(err)
            return Err(err)

        # This gets the forecast for this afternoon
        forecast_values = forecast_response["properties"]["periods"][0]

        return Ok(
            {
                "temperature": forecast_values["temperature"],
                "temperature_unit": forecast_values["temperatureUnit"],
                "short_forecast": forecast_values["shortForecast"],
            }
        )

File: golang/api/integrators/integrators.go
package integrators

import (
	"encoding/json"
	"errors"
	"fmt"
	"log"
)

type WeatherAPIClient struct {
	url string
}

func NewWeatherAPICient() WeatherAPIClient {
	log.Print("Starting up WeatherAPIClient")

	return WeatherAPIClient{url: "https://api.weather.gov"}
}

// Endpoints
func (wac WeatherAPIClient) GetForecast(
	longitude float32,
	latitude float32,
) (*Forecast, error) {
	/*
		Returns the most recent forecast data for today. We're first calling the
		'points' endpoint, which returns a url for the 'forecast' endpoint. We then
		call the 'forecast' endpoint to get the data we actually need.

		Points
		####################
		Documentation: https://www.weather.gov/documentation/services-web-api#/default/point
		Example:       https://api.weather.gov/points/39.7456,-97.0892

		Forecast
		####################
		Documentation: https://www.weather.gov/documentation/services-web-api#/default/gridpoint_forecast
		Example:       https://api.weather.gov/gridpoints/TOP/32,81/forecast
	*/

	// Points endpoint
	pointsUrl := fmt.Sprintf("%v/points/%v,%v", wac.url, longitude, latitude)
	pointsRespData, err := sendRequest(pointsUrl)
	if err != nil {
		log.Printf("error calling points endpoint | err=%v", err)
		return nil, err
	}

	var pointsObj PointsResponse
	json.Unmarshal(pointsRespData, &pointsObj)
	if pointsObj.Error != nil {
		errMsg := *pointsObj.Error
		log.Printf("points response has an error | err=%v", errMsg)
		return nil, errors.New(errMsg)
	}

	// Forecast endpoint
	forecastUrl := pointsObj.Properties.Forecast
	forecastRespData, err := sendRequest(forecastUrl)
	if err != nil {
		log.Printf("error calling forecast endpoint | err=%v", err)
		return nil, err
	}

	var forecastObj ForecastResponse
	json.Unmarshal(forecastRespData, &forecastObj)
	if forecastObj.Error != nil {
		errMsg := *forecastObj.Error
		log.Printf("forecast response has an error | err=%v", errMsg)
		return nil, errors.New(errMsg)
	}

	// This gets the most recent forecast for today
	forecastValues := forecastObj.Properties.Periods[0]

	return &Forecast{
		Temperature:     forecastValues.Temperature,
		TemperatureUnit: forecastValues.TemperatureUnit,
		ShortForecast:   forecastValues.ShortForecast,
	}, nil
}
